doc/fx.1
.\"t
.\" Automatically generated by Pandoc 2.2.3.2
.\"
.TH "fx" "1" "" "Version 20.0.2" "Command\-line JSON processing tool"
.hy
.SH NAME
.PP
\f[B]fx\f[] \[em] Command\-line JSON processing tool
.SH SYNOPSIS
.PP
\f[B]fx\f[] \f[B][options]\f[] \f[I]inputfile.json\f[]
.SH DESCRIPTION
.PP
fx can work in two modes: cli and interactive.
To start interactive mode pipe any JSON into fx:
.PP
$ curl \&... | fx
.PP
Or pass a filename as the first parameter:
.PP
$ fx my.json
.PP
If any argument was passed, fx will apply it and prints to stdout.
.SS Anonymous function
.PP
Use an anonymous function as reducer which gets JSON and processes it:
.IP
.nf
\f[C]
$\ echo\ \[aq]{"foo":\ [{"bar":\ "value"}]}\[aq]\ |\ fx\ \[aq]x\ =>\ x.foo[0].bar\[aq]
value
\f[]
.fi
.SS Binding
.PP
If you don't pass anonymous function \f[C]param\ =>\ ...\f[], code will
be automatically transformed into anonymous function.
And you can get access to JSON by \f[C]this\f[] keyword:
.IP
.nf
\f[C]
$\ echo\ \[aq]{"foo":\ [{"bar":\ "value"}]}\[aq]\ |\ fx\ \[aq]this.foo[0].bar\[aq]
value
\f[]
.fi
.SS Dot
.PP
It is possible to omit \f[C]this\f[] keyword:
.IP
.nf
\f[C]
$\ echo\ \[aq]{"foo":\ [{"bar":\ "value"}]}\[aq]\ |\ fx\ .foo[0].bar
value
\f[]
.fi
.PP
If a single dot is passed, the input JSON will be formatted but
otherwise unaltered:
.IP
.nf
\f[C]
$\ echo\ \[aq]{"foo":\ "bar"}\[aq]\ |\ fx\ .
{
\ \ "foo":\ "bar"
}
\f[]
.fi
.SS Map
.PP
One of the frequent operations is mapping some function on an array.
For example, to extract some values.
.IP
.nf
\f[C]
[
\ \ {
\ \ \ \ "author":\ {
\ \ \ \ \ \ "name":\ "antonmedv"
\ \ \ \ }
\ \ },
\ \ {...},
\ \ ...
]
\f[]
.fi
.PP
And we want to collect names of each object in the array.
We can do this by mapping anonymous function:
.IP
.nf
\f[C]
$\ cat\ ...\ |\ fx\ \[aq].map(x\ =>\ x.author.name)\[aq]
\f[]
.fi
.PP
Or we can do the same by using jq\-like syntax:
.IP
.nf
\f[C]
$\ cat\ ...\ |\ fx\ .[].author.name
[
\ \ "antonmedv",
\ \ ...
]
\f[]
.fi
.RS
.PP
Note what \f[C][]\f[] can be applied to map object values.
.IP
.nf
\f[C]
$\ echo\ \[aq]{"foo":\ 1,\ "bar":\ 2}\[aq]\ |\ fx\ .[]
[1,\ 2]
\f[]
.fi
.RE
.SS Chaining
.PP
You can pass any number of anonymous functions for reducing JSON:
.IP
.nf
\f[C]
$\ echo\ \[aq]{"foo":\ [{"bar":\ "value"}]}\[aq]\ |\ fx\ \[aq]x\ =>\ x.foo\[aq]\ \[aq]this[0]\[aq]\ \[aq]this.bar\[aq]
value
\f[]
.fi
.SS Updating
.PP
You can update existing JSON using the spread operator:
.IP
.nf
\f[C]
$\ echo\ \[aq]{"count":\ 0}\[aq]\ |\ fx\ \[aq]{...this,\ count:\ 1}\[aq]
{
\ \ "count":\ 1
}
\f[]
.fi
.SS Edit\-in\-place
.PP
\f[C]fx\f[] provides a function \f[C]save\f[] which will save everything
in place and return saved object.
This function can be only used with filename as first argument to
\f[C]fx\f[] command.
.PP
Usage:
.IP
.nf
\f[C]
fx\ data.json\ \[aq]{...this,\ count:\ this.count+1}\[aq]\ save\ .count
\f[]
.fi
.SS Query language
.PP
If you want to use query language, for example
jsonata (http://jsonata.org/) you can use helper function like this:
.IP
.nf
\f[C]
global.jsonata\ =\ expr\ =>\ require(\[aq]jsonata\[aq])(expr).evaluate
\f[]
.fi
.PP
And use it like this:
.IP
.nf
\f[C]
curl\ ...\ |\ fx\ \[aq]jsonata("$sum(Order.Product.(Price\ *\ Quantity))")\[aq]
\f[]
.fi
.PP
Instead you can create next alias in \f[I].bashrc\f[] file:
.IP
.nf
\f[C]
alias\ jsonata=\[aq]FX_APPLY=jsonata\ fx\[aq]
\f[]
.fi
.PP
And now all code arguments to \f[C]jsonata\f[] will be passed through
\f[C]jsonata\f[] helper.
And now you can use it like this:
.IP
.nf
\f[C]
curl\ ...\ |\ jsonata\ \[aq]$sum(Order.Product.(Price\ *\ Quantity))\[aq]
\f[]
.fi
.SS Formatting
.PP
If you need output other than JSON (for example arguments for xargs), do
not return anything from the reducer.
\f[C]undefined\f[] value is printed into stderr by default.
.IP
.nf
\f[C]
echo\ \[aq][]\[aq]\ |\ fx\ \[aq]void\ 0\[aq]
undefined
\f[]
.fi
.IP
.nf
\f[C]
echo\ \[aq][1,2,3]\[aq]\ |\ fx\ \[aq]this.forEach(x\ =>\ console.log(+x))\[aq]\ 2>/dev/null\ |\ xargs\ echo
1\ 2\ 3
\f[]
.fi
.SS Other examples
.PP
Convert object to array:
.IP
.nf
\f[C]
$\ cat\ package.json\ |\ fx\ \[aq]Object.keys(this.dependencies)\[aq]
\f[]
.fi
.PP
Or by two functions:
.IP
.nf
\f[C]
$\ cat\ package.json\ |\ fx\ .dependencies\ Object.keys
\f[]
.fi
.PP
By the way, fx has shortcut for \f[C]Object.keys\f[].
Previous example can be rewritten as:
.IP
.nf
\f[C]
$\ cat\ package.json\ |\ fx\ .dependencies\ ?
\f[]
.fi
.SS Streaming mode
.PP
\f[C]fx\f[] supports line\-delimited JSON and concatenated JSON
streaming.
.IP
.nf
\f[C]
$\ kubectl\ logs\ ...\ |\ fx\ .message
\f[]
.fi
.RS
.PP
Note what is object lacks \f[C]message\f[] field, \f[I]undefined\f[]
will be printed to stderr.
This is useful to see if you are skipping some objects.
But if you want to hide them, redirect stderr to \f[C]/dev/null\f[].
.RE
.SS Filtering
.PP
Sometimes it is necessary to omit some messages in JSON stream, or
select only specified log messages.
For this purpose, \f[C]fx\f[] has special helpers
\f[C]select\f[]/\f[C]filter\f[], pass function into it to select/filter
JSON messages.
.IP
.nf
\f[C]
$\ kubectl\ logs\ ...\ |\ fx\ \[aq]select(x\ =>\ x.status\ ==\ 500)\[aq]\ .message
\f[]
.fi
.IP
.nf
\f[C]
$\ kubectl\ logs\ ...\ |\ fx\ \[aq]filter(x\ =>\ x.status\ <\ 499)\[aq]\ .message
\f[]
.fi
.RS
.PP
If \f[C]filter\f[]/\f[C]select\f[] overridden in \f[I].fxrc\f[] you
still able to access them with prefix: \f[C]std.select(cb)\f[] or
\f[C]std.filter(cd)\f[]
.RE
.SS Interactive mode
.PP
Click on fields to expand or collapse JSON tree, use mouse wheel to
scroll view.
.PP
Next commands available in interactive mode:
.PP
.TS
tab(@);
l l.
T{
Key
T}@T{
Command
T}
_
T{
\f[C]q\f[] or \f[C]Esc\f[] or \f[C]Ctrl\f[]+\f[C]c\f[]
T}@T{
Exit
T}
T{
\f[C]up\f[] or \f[C]k\f[]
T}@T{
Move cursor up
T}
T{
\f[C]down\f[] or \f[C]j\f[]
T}@T{
Move cursor down
T}
T{
\f[C]left\f[] or \f[C]h\f[]
T}@T{
Collapse
T}
T{
\f[C]right\f[] or \f[C]l\f[]
T}@T{
Expand
T}
T{
\f[C]Shift\f[]+\f[C]right\f[] or \f[C]L\f[]
T}@T{
Expand all under cursor
T}
T{
\f[C]Shift\f[]+\f[C]left\f[] or \f[C]K\f[]
T}@T{
Collapse all under cursor
T}
T{
\f[C]e\f[]
T}@T{
Expand all
T}
T{
\f[C]E\f[]
T}@T{
Collapse all
T}
T{
\f[C]g\f[]
T}@T{
Scroll to top
T}
T{
\f[C]G\f[]
T}@T{
Scroll to bottom
T}
T{
\f[C]\&.\f[]
T}@T{
Edit filter
T}
T{
\f[C]/\f[]
T}@T{
Search
T}
T{
\f[C]n\f[]
T}@T{
Find next
T}
T{
\f[C]p\f[]
T}@T{
Exit and print JSON to stdout
T}
T{
\f[C]P\f[]
T}@T{
Exit and print fully expanded JSON to stdout
T}
.TE
.PP
These commands are available when editing the filter:
.PP
.TS
tab(@);
l l.
T{
Key
T}@T{
Command
T}
_
T{
\f[C]Enter\f[]
T}@T{
Apply filter
T}
T{
\f[C]Ctrl\f[]+\f[C]u\f[]
T}@T{
Clear filter
T}
T{
\f[C]Ctrl\f[]+\f[C]w\f[]
T}@T{
Delete last part
T}
T{
\f[C]up\f[]/\f[C]down\f[]
T}@T{
Select autocomplete
T}
.TE
.SS Searching
.PP
Press \f[C]/\f[] and type regexp pattern to search in current JSON.
Search work with currently applied filter.
.PP
Examples of pattern and corresponding regexp:
.PP
.TS
tab(@);
l l.
T{
Pattern
T}@T{
RegExp
T}
_
T{
\f[C]/apple\f[]
T}@T{
\f[C]/apple/ig\f[]
T}
T{
\f[C]/apple/\f[]
T}@T{
\f[C]/apple/\f[]
T}
T{
\f[C]/apple/u\f[]
T}@T{
\f[C]/apple/u\f[]
T}
T{
\f[C]/\\w+\f[]
T}@T{
\f[C]/\\w+/ig\f[]
T}
.TE
.SS Selecting text
.PP
You may found what you can't just select text in fx.
This is due the fact that all mouse events redirected to stdin.
To be able select again you need instruct your terminal not to do it.
This can be done by holding special keys while selecting:
.PP
.TS
tab(@);
l l.
T{
Key
T}@T{
Terminal
T}
_
T{
\f[C]Option\f[]+\f[C]Mouse\f[]
T}@T{
iTerm2, Hyper
T}
T{
\f[C]Fn\f[]+\f[C]Mouse\f[]
T}@T{
Terminal.app
T}
T{
\f[C]Shift\f[]+\f[C]Mouse\f[]
T}@T{
Linux
T}
.TE
.RS
.PP
Note what you can press \f[C]p\f[]/\f[C]P\f[] to print everything to
stdout and select if there.
.RE
.SS Options
.TP
.B \[en]version
Show program version
.RS
.RE
.TP
.B \[en]help
Display help information
.RS
.RE
.SH FILES
.SS Using packages
.PP
Use any npm package by installing it globally:
.IP
.nf
\f[C]
$\ npm\ install\ \-g\ lodash
$\ cat\ package.json\ |\ fx\ \[aq]require("lodash").keys(this.dependencies)\[aq]
\f[]
.fi
.SS Using .fxrc
.PP
Create \f[I].fxrc\f[] file in \f[C]$HOME\f[] directory, and require any
packages or define global functions.
.PP
For example, access all lodash methods without \f[C]_\f[] prefix.
Put in your \f[C]\&.fxrc\f[] file:
.IP
.nf
\f[C]
Object.assign(global,\ require(\[aq]lodash/fp\[aq]))
\f[]
.fi
.PP
And now you will be able to call all lodash methods.
For example, see who's been committing to react recently:
.IP
.nf
\f[C]
curl\ \[aq]https://api.github.com/repos/facebook/react/commits?per_page=100\[aq]\ \\
|\ fx\ \[aq]groupBy("commit.author.name")\[aq]\ \[aq]mapValues(size)\[aq]\ toPairs\ \[aq]sortBy(1)\[aq]\ reverse\ \[aq]take(10)\[aq]\ fromPairs
\f[]
.fi
.RS
.PP
To be able require global modules make sure you have correct
\f[C]NODE_PATH\f[] env variable.
.IP
.nf
\f[C]
export\ NODE_PATH=`npm\ root\ \-g`
\f[]
.fi
.RE
.SH ENVIRONMENT
.PP
You may find that sometimes, on really big JSON files, fx prints an
error message like this:
.IP
.nf
\f[C]
FATAL\ ERROR:\ JavaScript\ heap\ out\ of\ memory\ 
\f[]
.fi
.PP
V8 limits memory usage to around 2 GB by default.
You can increase the limit by putting this line in your
\f[I].profile\f[]:
.IP
.nf
\f[C]
export\ NODE_OPTIONS=\[aq]\-\-max\-old\-space\-size=8192\[aq]
\f[]
.fi
.SH BUGS
.PP
See GitHub Issues: <https://github.com/antonmedv/fx/issues>
.SH AUTHOR
.PP
Anton Medvedev (anton+github\@medv.io)
.PP
<https://github.com/antonmedv/fx>


fx.js
'use strict'
const fs = require('fs')
const tty = require('tty')
const blessed = require('@medv/blessed')
const stringWidth = require('string-width')
const reduce = require('./reduce')
const print = require('./print')
const find = require('./find')
const config = require('./config')

module.exports = function start(filename, source, prev = {}) {
  // Current rendered object on a screen.
  let json = prev.json || source

  // Contains map from row number to expand path.
  // Example: {0: '', 1: '.foo', 2: '.foo[0]'}
  let index = new Map()

  // Contains expanded paths. Example: ['', '.foo']
  // Empty string represents root path.
  const expanded = prev.expanded || new Set()
  expanded.add('')

  // Current filter code.
  let currentCode = null

  // Current search regexp and generator.
  let highlight = null
  let findGen = null
  let currentPath = null

  let ttyReadStream, ttyWriteStream

  // Reopen tty
  if (process.platform === 'win32') {
    const cfs = process.binding('fs')
    ttyReadStream = tty.ReadStream(cfs.open('conin$', fs.constants.O_RDWR | fs.constants.O_EXCL, 0o666))
    ttyWriteStream = tty.WriteStream(cfs.open('conout$', fs.constants.O_RDWR | fs.constants.O_EXCL, 0o666))
  } else {
    const ttyFd = fs.openSync('/dev/tty', 'r+')
    ttyReadStream = tty.ReadStream(ttyFd)
    ttyWriteStream = tty.WriteStream(ttyFd)
  }

  const program = blessed.program({
    input: ttyReadStream,
    output: ttyWriteStream,
  })

  const screen = blessed.screen({
    program: program,
    smartCSR: true,
    fullUnicode: true,
  })

  const box_height = config.useRuler ? '100%-2' : '100%'

  const box = blessed.box({
    parent: screen,
    tags: false,
    left: 0,
    top: 0,
    width: '100%',
    height: box_height,
    mouse: true,
    keys: true,
    vi: true,
    ignoreArrows: true,
    alwaysScroll: true,
    scrollable: true,
  })

  const input = blessed.textbox({
    parent: screen,
    bottom: 0,
    left: 0,
    height: 1,
    width: '100%',
  })

  const search = blessed.textbox({
    parent: screen,
    bottom: 0,
    left: 0,
    height: 1,
    width: '100%',
  })

  const statusBar = blessed.box({
    parent: screen,
    tags: false,
    bottom: 0,
    left: 0,
    height: 1,
    width: '100%',
  })

  const ruler = config.useRuler && blessed.box({
    parent: screen,
    fg: config.ruler.fg,
    bg: config.ruler.bg,
    tags: false,
    align: 'right',
    bottom: 1,
    right: 0,
    height: 1,
    width: 16,
  })

  const pathBar = config.useRuler && blessed.box({
    parent: screen,
    fg: config.ruler.path,
    bg: config.ruler.bg,
    tags: false,
    align: 'left',
    bottom: 1,
    left: 0,
    height: 1,
    width: '100%-16',
  })

  const autocomplete = blessed.list({
    parent: screen,
    width: 6,
    height: 7,
    left: 1,
    bottom: 1,
    style: config.list,
  })

  screen.title = filename
  box.focus()
  input.hide()
  search.hide()
  statusBar.hide()
  autocomplete.hide()

  process.stdout.on('resize', () => {
    // Blessed has a bug with resizing the terminal. I tried my best to fix it but was not succeeded.
    // For now exit and print seem like a reasonable alternative, as it not usable after resize.
    // If anyone can fix this bug it will be cool.
    printJson({expanded})
  })

  screen.key(['escape', 'q', 'C-c'], function () {
    exit()
  })

  input.on('submit', function () {
    if (autocomplete.hidden) {
      const code = input.getValue()
      if (/^\//.test(code)) {
        // Forgive a mistake to the user. This looks like user wanted to search something.
        apply('')
        applyPattern(code)
      } else {
        apply(code)
      }
    } else {
      // Autocomplete selected
      let code = input.getValue()
      let replace = autocomplete.getSelected()
      if (/^[a-z]\w*$/i.test(replace)) {
        replace = '.' + replace
      } else {
        replace = `["${replace}"]`
      }
      code = code.replace(/\.\w*$/, replace)

      input.setValue(code)
      autocomplete.hide()
      update(code)

      // Keep editing code
      input.readInput()
    }
  })

  input.on('cancel', function () {
    if (autocomplete.hidden) {
      const code = input.getValue()
      apply(code)
    } else {
      // Autocomplete not selected
      autocomplete.hide()
      render()

      // Keep editing code
      input.readInput()
    }
  })

  input.on('update', function (code) {
    if (currentCode === code) {
      return
    }
    currentCode = code
    if (index.size < 10000) { // Don't live update in we have a big JSON file.
      update(code)
    }
    complete(code)
  })

  input.key('up', function () {
    if (!autocomplete.hidden) {
      autocomplete.up()
      render()
    }
  })

  input.key('down', function () {
    if (!autocomplete.hidden) {
      autocomplete.down()
      render()
    }
  })

  input.key('C-u', function () {
    input.setValue('')
    update('')
    render()
  })

  input.key('C-w', function () {
    let code = input.getValue()
    code = code.replace(/[\.\[][^\.\[]*$/, '')
    input.setValue(code)
    update(code)
    render()
  })

  search.on('submit', function (pattern) {
    applyPattern(pattern)
  })

  search.on('cancel', function () {
    highlight = null
    currentPath = null

    search.hide()
    search.setValue('')

    box.height = box_height
    box.focus()

    program.cursorPos(0, 0)
    render()
  })

  box.key('.', function () {
    hideStatusBar()
    box.height = config.useRuler ? box_height : '100%-1'
    input.show()
    if (input.getValue() === '') {
      input.setValue('.')
      complete('.')
    }
    input.readInput()
    render()
  })

  box.key('/', function () {
    hideStatusBar()
    box.height = config.useRuler ? box_height : '100%-1'
    search.show()
    search.setValue('/')
    search.readInput()
    render()
  })

  box.key('e', function () {
    hideStatusBar()
    expanded.clear()
    for (let path of dfs(json)) {
      if (expanded.size < 1000) {
        expanded.add(path)
      } else {
        break
      }
    }
    render()
  })

  box.key('S-e', function () {
    hideStatusBar()
    expanded.clear()
    expanded.add('')
    render()

    // Make sure cursor stay on JSON object.
    const [n] = getLine(program.y)
    if (typeof n === 'undefined' || !index.has(n)) {
      // No line under cursor
      let rest = [...index.keys()]
      if (rest.length > 0) {
        const next = Math.max(...rest)
        let y = box.getScreenNumber(next) - box.childBase
        if (y <= 0) {
          y = 0
        }
        const line = box.getScreenLine(y + box.childBase)
        program.cursorPos(y, line.search(/\S/))
      }
    }
  })

  box.key('n', function () {
    hideStatusBar()
    findNext()
  })

  // High, Middle, Low cursor movement
  box.key('S-h', function () {
    hideStatusBar()
    program.showCursor()
    const line = box.getScreenLine(box.childBase)
    program.cursorPos(0, line.search(/\S/))
  })

  box.key('S-m', function () {
    hideStatusBar()
    program.showCursor()
    const lastLine = box.height < box.getScrollHeight() ? box.height : box.getScrollHeight()
    const offset = lastLine / 2
    const line = box.getScreenLine(box.childBase + offset)
    program.cursorPos(offset, line.search(/\S/))
  })

  box.key('S-l', function () {
    hideStatusBar()
    program.showCursor()
    const lastLine = box.height < box.getScrollHeight() ? box.height : box.getScrollHeight()
    const line = box.getScreenLine(box.childBase + lastLine - 1)
    program.cursorPos(lastLine - 1, line.search(/\S/))
  })


  // Scrolls to and sets cursor at first line of object
  box.key('g', function () {
    hideStatusBar()
    program.showCursor()
    box.scrollTo(0)

    const line = box.getScreenLine(0)
    program.cursorPos(0)
    render()
  })

  // Scrolls to and sets cursor on last line of object
  box.key('S-g', function () {
    const lastLine = box.getScrollHeight() - 1

    hideStatusBar()
    program.showCursor()
    box.scrollTo(lastLine)

    program.cursorPos(box.height < box.getScrollHeight() ? box.height - 1 : lastLine)
    render()
  })

  box.key(['up', 'k'], function () {
    hideStatusBar()
    program.showCursor()
    const [n] = getLine(program.y)

    let next
    for (let [i,] of index) {
      if (i < n && (typeof next === 'undefined' || i > next)) {
        next = i
      }
    }

    if (typeof next !== 'undefined') {
      let y = box.getScreenNumber(next) - box.childBase
      if (y <= 0) {
        box.scroll(-1)
        y = 0
      }

      const line = box.getScreenLine(y + box.childBase)
      program.cursorPos(y, line.search(/\S/))
    }
    render()
  })

  // Half page up
  box.key(['C-u','u'], function () {
    hideStatusBar()
    program.showCursor()
    const page = Math.round(box.height / 2)

    box.scroll(-page || -1)

    let y = program.y
    if (box.getScroll() == 0) {
      y -= page
    } else {
      y = box.height - 1
    }

    if (y < 0) {
      y = 0
    }

    const line = box.getScreenLine(y + box.childBase)
    program.cursorPos(y, line.search(/\S/))
    render()
  })

  // Full page up (backwards)
  box.key(['C-b','b','pageup'], function () {
    hideStatusBar()
    program.showCursor()
    box.scroll(-box.height || -1)

    let y = box.height - 1
    if (box.getScroll() < box.height) {
      y -= box.height
    }
    if (y < 0) {
      y = 0
    }

    const line = box.getScreenLine(y + box.childBase)
    program.cursorPos(y, line.search(/\S/))
    render()
  })

  box.key(['down', 'j'], function () {
    hideStatusBar()
    program.showCursor()
    const [n] = getLine(program.y)

    let next
    for (let [i,] of index) {
      if (i > n && (typeof next === 'undefined' || i < next)) {
        next = i
      }
    }

    if (typeof next !== 'undefined') {
      let y = box.getScreenNumber(next) - box.childBase
      if (y >= box.height) {
        box.scroll(1)
        y = box.height - 1
      }

      const line = box.getScreenLine(y + box.childBase)
      program.cursorPos(y, line.search(/\S/))
    }
    render()
  })

  // Half page down
  box.key(['C-d','d'], function () {
    hideStatusBar()
    program.showCursor()
    const page = Math.floor(box.height / 2)
    const lastLine = box.getScrollHeight()

    let y = program.y
    if(box.childBase + page < lastLine - page) {
      box.scroll(page)
      y = 0
    } else if (box.height < lastLine) {
      box.scroll(page)
      if (y + page > box.height) {
        y = box.height - 1
      } else {
        y += page
        if(y >= box.height) {
          y = box.height - 1
        }
      }
    } else {
      y = lastLine - 1
    }

    const line = box.getScreenLine(y + box.childBase)
    program.cursorPos(y, line && line.search(/\S/))
    render()
  })

  // Full page down (forwards)
  box.key(['C-f','f','pagedown'], function () {
    hideStatusBar()
    program.showCursor()
    const lastLine = box.getScrollHeight()

    let y = program.y
    if(box.childBase + box.height < lastLine - box.height) {
      box.scroll(box.height)
      y = 0
    } else if (box.height < lastLine) {
      box.scroll(box.height)
      y = box.height - 1
    } else {
      y = lastLine - 1
    }

    const line = box.getScreenLine(y + box.childBase)
    program.cursorPos(y, line.search(/\S/))
    render()
  })

  const isExpanded = y => expanded.has(index.get(y + box.childBase))

  // Next expanded object/array
  box.key('}', function() {
    hideStatusBar()
    const [n] = getLine(program.y)
    if(n < box.getScrollHeight() - 1) {
      let y = program.y

      program.showCursor()
      do {
        y += 1
        if(y >= box.height) {
          box.scroll(box.height || 1)
          // don't jump cursor to top when scrolling to patial last page
          y = n < box.getScrollHeight() - box.height ? 0 : n % box.height
        }
      } while(!isExpanded(y) && y + box.childBase < box.getScrollHeight() - 1)

      const line = box.getScreenLine(y + box.childBase)
      program.cursorPos(y, line && line.search(/\S/))
      render()
    }
  })

  // Previous expanded object/array
  box.key('{', function() {
    hideStatusBar()
    const [n] = getLine(program.y)
    if(box.childBase > 0 || program.y > 0) {
      let y = program.y

      program.showCursor()
      do {
        y -= 1
        if(y < 0) {
          box.scroll(-box.height || -1)
          // don't jump cursor to bottom when scrolling to patial first page
          y = n < box.height ? n - 1 : box.height
        }
      } while(!isExpanded(y) && n > 1)

      const line = box.getScreenLine(y + box.childBase)
      program.cursorPos(y, line && line.search(/\S/))
      render()
    }
  })

  box.key(['right', 'l'], function () {
    hideStatusBar()
    const [n, line] = getLine(program.y)
    program.showCursor()
    program.cursorPos(program.y, line.search(/\S/))
    const path = index.get(n)
    if (!expanded.has(path)) {
      expanded.add(path)
      render()
    }
  })

  // Expand everything under cursor.
  box.key(['S-right','S-o'], function () {
    hideStatusBar()
    const [n, line] = getLine(program.y)
    program.showCursor()
    program.cursorPos(program.y, line.search(/\S/))
    const path = index.get(n)
    const subJson = reduce(json, 'this' + path)
    for (let p of dfs(subJson, path)) {
      if (expanded.size < 1000) {
        expanded.add(p)
      } else {
        break
      }
    }
    render()
  })

  box.key(['S-left', 'S-k'], function () {
    hideStatusBar()
    const [n, line] = getLine(program.y)
    program.showCursor()
    program.cursorPos(program.y, line.search(/\S/))
    const path = index.get(n)
    if (!path) {
      // collapse on top level (should render like after run)
      expanded.clear()
      expanded.add('')
    } else {
      const subJson = reduce(json, 'this' + path)
      for (let p of dfs(subJson, path)) {
        if (expanded.size < 1000) {
          expanded.delete(p)
        } else {
          break
        }
      }
    }
    render()
  })

  box.key(['left', 'h'], function () {
    hideStatusBar()
    const [n, line] = getLine(program.y)
    program.showCursor()
    program.cursorPos(program.y, line.search(/\S/))

    // Find path at current cursor position.
    const path = index.get(n)

    if (expanded.has(path)) {
      // Collapse current path.
      expanded.delete(path)
      render()
    } else {
      // If there is no expanded paths on current line,
      // collapse parent path of current location.
      if (typeof path === 'string') {
        // Trip last part (".foo", "[0]") to get parent path.
        const parentPath = path.replace(/(\.[^\[\].]+|\[\d+\])$/, '')
        if (expanded.has(parentPath)) {
          expanded.delete(parentPath)
          render()

          // Find line number of parent path, and if we able to find it,
          // move cursor to this position of just collapsed parent path.
          for (let y = program.y; y >= 0; --y) {
            const [n, line] = getLine(y)
            if (index.get(n) === parentPath) {
              program.cursorPos(y, line.search(/\S/))
              break
            }
          }
        }
      }
    }
  })

  box.on('click', function (mouse) {
    hideStatusBar()
    const [n, line] = getLine(mouse.y)
    if (mouse.x >= stringWidth(line)) {
      return
    }

    program.hideCursor()
    program.cursorPos(mouse.y, line.search(/\S/))
    autocomplete.hide()

    const path = index.get(n)
    if (expanded.has(path)) {
      expanded.delete(path)
    } else {
      expanded.add(path)
    }
    render()
  })

  box.on('scroll', function () {
    hideStatusBar()
  })

  box.key('p', function () {
    printJson({expanded})
  })

  box.key('S-p', function () {
    printJson()
  })

  function printJson(options = {}) {
    screen.destroy()
    program.disableMouse()
    program.destroy()
    setTimeout(() => {
      const [text] = print(json, options)
      console.log(text)
      process.exit(0)
    }, 10)
  }

  function getLine(y) {
    const dy = box.childBase + y
    const n = box.getNumber(dy)
    const line = box.getScreenLine(dy)
    if (typeof line === 'undefined') {
      return [n, '']
    }
    return [n, line]
  }

  function apply(code) {
    if (code && code.length !== 0) {
      try {
        json = reduce(source, code)
      } catch (e) {
        // pass
      }
    } else {
      box.height = box_height
      input.hide()
      json = source
    }
    box.focus()
    program.cursorPos(0, 0)
    render()
  }

  function complete(inputCode) {
    const match = inputCode.match(/\.(\w*)$/)
    const code = /^\.\w*$/.test(inputCode) ? '.' : inputCode.replace(/\.\w*$/, '')

    let json
    try {
      json = reduce(source, code)
    } catch (e) {
    }

    if (match) {
      if (typeof json === 'object' && json.constructor === Object) {
        const keys = Object.keys(json)
          .filter(key => key.startsWith(match[1]))
          .slice(0, 1000) // With lots of items, list takes forever to render.

        // Hide if there is nothing to show or
        // don't show if there is complete match.
        if (keys.length === 0 || (keys.length === 1 && keys[0] === match[1])) {
          autocomplete.hide()
          return
        }

        autocomplete.width = Math.max(...keys.map(key => key.length)) + 1
        autocomplete.height = Math.min(7, keys.length)
        autocomplete.left = Math.min(
          screen.width - autocomplete.width,
          code.length === 1 ? 1 : code.length + 1
        )

        let selectFirst = autocomplete.items.length !== keys.length
        autocomplete.setItems(keys)

        if (selectFirst) {
          autocomplete.select(autocomplete.items.length - 1)
        }
        if (autocomplete.hidden) {
          autocomplete.show()
        }
      } else {
        autocomplete.clearItems()
        autocomplete.hide()
      }
    }
  }

  function update(code) {
    if (code && code.length !== 0) {
      try {
        const pretender = reduce(source, code)
        if (
          typeof pretender !== 'undefined'
          && typeof pretender !== 'function'
          && !(pretender instanceof RegExp)
        ) {
          json = pretender
        }
      } catch (e) {
        // pass
      }
    }
    if (code === '') {
      json = source
    }

    if (highlight) {
      findGen = find(json, highlight)
    }
    render()
  }

  function applyPattern(pattern) {
    let regex
    let m = pattern.match(/^\/(.*)\/([gimuy]*)$/)
    if (m) {
      try {
        regex = new RegExp(m[1], m[2])
      } catch (e) {
        // Wrong regexp.
      }
    } else {
      m = pattern.match(/^\/(.*)$/)
      if (m) {
        try {
          regex = new RegExp(m[1], 'gi')
        } catch (e) {
          // Wrong regexp.
        }
      }
    }
    highlight = regex

    search.hide()

    if (highlight) {
      findGen = find(json, highlight)
      findNext()
    } else {
      findGen = null
      currentPath = null
    }
    search.setValue('')

    box.height = box_height
    box.focus()

    program.cursorPos(0, 0)
    render()
  }

  function findNext() {
    if (!findGen) {
      return
    }

    const {value: path, done} = findGen.next()

    if (done) {
      showStatusBar('Pattern not found')
    } else {

      currentPath = ''
      for (let p of path) {
        expanded.add(currentPath += p)
      }
      render()

      for (let [k, v] of index) {
        if (v === currentPath) {
          let y = box.getScreenNumber(k)

          // Scroll one line up for better view and make sure it's not negative.
          if (--y < 0) {
            y = 0
          }

          box.scrollTo(y)
          render()
        }
      }

      // Set cursor to current path.
      // We need timeout here to give our terminal some time.
      // Without timeout first cursorPos call does not working,
      // it looks like an ugly hack and it is an ugly hack.
      setTimeout(() => {
        for (let [k, v] of index) {
          if (v === currentPath) {
            let y = box.getScreenNumber(k) - box.childBase
            if (y <= 0) {
              y = 0
            }
            const line = box.getScreenLine(y + box.childBase)
            program.cursorPos(y, line.search(/\S/))
            render()
          }
        }
      }, 100)
    }
  }

  function updateRuler() {
    const y = box.childBase + program.y
    const [n] = getLine(program.y)
    const path = index.get(n) || ''
    const scrollPercent = ((y / (box.getScrollHeight() - 1)) * 100).toFixed(0)
    const compressPath = p => {
      if(pathBar.strWidth(path) < pathBar.width) {
        return p
      } else {
        return '…' + p.substring(path.length - pathBar.width + 1)
      }
    }
    pathBar.show()
    pathBar.setContent(`${compressPath(path)}`)
    ruler.show()
    ruler.setContent(`${box.childBase + program.y}\t${scrollPercent}%`)
    screen.render()
  }

  function showStatusBar(status) {
    statusBar.show()
    statusBar.setContent(config.statusBar(` ${status} `))
    render()
  }

  function hideStatusBar() {
    if (!statusBar.hidden) {
      statusBar.hide()
      statusBar.setContent('')
      render()
    }
  }

  function render() {
    let content
    [content, index] = print(json, {expanded, highlight, currentPath})

    if (typeof content === 'undefined') {
      content = 'undefined'
    }

    box.setContent(content)
    config.useRuler && updateRuler() 
    screen.render()
  }

  function exit() {
    // If exit program immediately, stdin may still receive
    // mouse events which will be printed in stdout.
    program.disableMouse()
    setTimeout(() => process.exit(0), 10)
  }

  render()
}

function* bfs(json) {
  const queue = [[json, '']]

  while (queue.length > 0) {
    const [v, path] = queue.shift()

    if (!v) {
      continue
    }

    if (Array.isArray(v)) {
      yield path
      let i = 0
      for (let item of v) {
        const p = path + '[' + (i++) + ']'
        queue.push([item, p])
      }
    }

    if (typeof v === 'object' && v.constructor === Object) {
      yield path
      for (let [key, value] of Object.entries(v)) {
        const p = path + '.' + key
        queue.push([value, p])
      }
    }
  }
}

function* dfs(v, path = '') {
  if (!v) {
    return
  }

  if (Array.isArray(v)) {
    yield path
    let i = 0
    for (let item of v) {
      yield* dfs(item, path + '[' + (i++) + ']')
    }
  }

  if (typeof v === 'object' && v.constructor === Object) {
    yield path
    for (let [key, value] of Object.entries(v)) {
      yield* dfs(value, path + '.' + key)
    }
  }
}
